sdk/core/core-rest-pipeline/src/retryStrategies/failoverRetryStrategy.ts

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { PipelineRequest } from "../interfaces";
+import { RetryInformation, RetryModifiers, RetryStrategy } from "./retryStrategy";
+import { DEFAULT_CLIENT_MAX_RETRY_INTERVAL, DEFAULT_CLIENT_RETRY_INTERVAL } from "../constants";
+import { DefaultRetryPolicyOptions } from "../policies/defaultRetryPolicy";
+import { exponentialDelayInMs } from "../util/retryAfter";
+
+/**
+ * An iterator factory for failover hosts. Yields the next host to retry, or yields undefined if failover should be skipped.
+ */
+export type FailoverHostIteratorFactory = (
+  retryState: RetryInformation,
+  options: DefaultRetryPolicyOptions
+) => Iterator<FailoverHostState | undefined, void, RetryInformation>;
+
+/**
+ * A host for failover and its associated state.
+ */
+export interface FailoverHostState {
+  /**
+   * The host which this policy will try next.
+   */
+  host: string;
+  /**
+   * How many times this host has been tried.
+   */
+  retryCount: number;
+  /**
+   * The policy should wait until `retryAt` has elapsed to try this host.
+   */
+  retryAt?: Date;
+}
+interface FailoverHosts {
+  index: number;
+  hostStates: FailoverHostState[];
+}
+
+/**
+ * Default behavior for generating hosts for failover.
+ *
+ * If the request method is GET, HEAD, or OPTIONS, it cycles through `readHosts`. Otherwise, it cycles through `writeHosts`.
+ */
+export function readWriteFailoverHostIteratorFactory(options: {
+  readHosts?: string[];
+  writeHosts?: string[];
+}): FailoverHostIteratorFactory {
+  const { readHosts, writeHosts } = options;
```

Comment:
another place we could use return is if both readHosts and writeHosts are undefined or empty arrays

Reply:
Good catch! I added an optimization for this case - the returned generator now immediately returns when both arrays are empty.